platforms/macos/src/subclass.rs

```diff
@@ -0,0 +1,183 @@
+    declare_class,
+    ffi::{
+        objc_getAssociatedObject, objc_setAssociatedObject, object_setClass,
+        OBJC_ASSOCIATION_ASSIGN,
+    },
+    foundation::{NSArray, NSObject, NSPoint},
+    msg_send_id,
+    rc::{Id, Owned, WeakId},
+    runtime::{Class, Sel},
+    sel, ClassType,
+};
+use once_cell::sync::Lazy;
+use parking_lot::Mutex;
+use std::{collections::HashMap, ffi::c_void, ops::Deref};
+
+use crate::{appkit::NSView, Adapter};
+
+static SUBCLASSES: Lazy<Mutex<HashMap<&'static Class, &'static Class>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+// Declare as mutable to ensure the address is unique.
+static mut ASSOCIATED_OBJECT_KEY: u8 = 0;
+
+fn associated_object_key() -> *const c_void {
+    unsafe { &ASSOCIATED_OBJECT_KEY as *const u8 as *const _ }
+}
+
+declare_class!(
+    struct AssociatedObject {
+        // SAFETY: These are set in AssociatedObject::new, immediately after
+        // the object is created.
+        adapter: IvarDrop<Box<Adapter>>,
+        prev_class: &'static Class,
+    }
+
+    unsafe impl ClassType for AssociatedObject {
+        type Super = NSObject;
+        const NAME: &'static str = "AccessKitSubclassAssociatedObject";
+    }
+);
+
+impl AssociatedObject {
+    fn new(adapter: Adapter, prev_class: &'static Class) -> Id<Self, Owned> {
+        unsafe {
+            let mut object: Id<Self, Owned> = msg_send_id![Self::class(), new];
+            Ivar::write(&mut object.adapter, Box::new(adapter));
+            Ivar::write(&mut object.prev_class, prev_class);
+            object
+        }
+    }
+}
+
+fn associated_object(view: &NSView) -> &AssociatedObject {
+    unsafe {
+        (objc_getAssociatedObject(view as *const NSView as *const _, associated_object_key())
+            as *const AssociatedObject)
+            .as_ref()
+    }
+    .unwrap()
+}
+
+// Some view classes, like the one in winit 0.27, assume that they are the
+// lowest subclass, and call [self superclass] to get their superclass.
+// Give them the answer they need.
+unsafe extern "C" fn superclass(this: &NSView, _cmd: Sel) -> Option<&Class> {
+    let associated = associated_object(this);
+    associated.prev_class.superclass()
+}
+
+unsafe extern "C" fn children(this: &NSView, _cmd: Sel) -> *mut NSArray<NSObject> {
+    let associated = associated_object(this);
+    associated.adapter.view_children()
+}
+
+unsafe extern "C" fn focus(this: &NSView, _cmd: Sel) -> *mut NSObject {
+    let associated = associated_object(this);
+    associated.adapter.focus()
+}
+
+unsafe extern "C" fn hit_test(this: &NSView, _cmd: Sel, point: NSPoint) -> *mut NSObject {
+    let associated = associated_object(this);
+    associated.adapter.hit_test(point)
+}
+
+/// Uses dynamic Objective-C subclassing to implement the NSView
+/// accessibility methods when normal subclassing isn't an option.
+pub struct SubclassingAdapter {
+    view: WeakId<NSView>,
+    associated: Id<AssociatedObject, Owned>,
```

Comment:
After you've set the associated object, this is no longer the only pointer to the object (and hence `Owned` is wrong)
```suggestion
    associated: Id<AssociatedObject, Shared>,
```